cpp/test/sg/fil_test.cu

```diff
@@ -886,4 +889,139 @@ TEST_P(TreeliteThrowSparse8FilTest, Import) { check(); }
 INSTANTIATE_TEST_CASE_P(FilTests, TreeliteThrowSparse8FilTest,
                         testing::ValuesIn(import_throw_sparse8_inputs));
 
+template <typename T>
+__device__ void serial_multi_reduction(T* in, T* out, int set_size,
+                                       int n_sets) {
+  __syncthreads();
+  if (threadIdx.x < set_size) {
+    int reduction_id = threadIdx.x;
+    T sum = 0;
+    for (int set = 0; set < n_sets; ++set)
+      sum += in[reduction_id + set * set_size];
+    out[reduction_id] = sum;
+  }
+  __syncthreads();
+}
+
+// the most threads a block can have
+const int max_threads = 1024;
+
+struct MultiReductionTestParams {
+  int radix, set_size, n_sets;
+};
+
+template <int R, typename T>
+__device__ void test_single_radix(T value, MultiReductionTestParams p,
+                                  int& error) {
+  __shared__ T work[max_threads], correct_result[max_threads];
+  work[threadIdx.x] = value;
+  serial_multi_reduction(work, correct_result, p.set_size, p.n_sets);
+  T sum = multi_reduction<R>(work, p.set_size, p.n_sets);
+  if (threadIdx.x < p.set_size &&
+      1e-4 < fabsf(sum - correct_result[threadIdx.x])) {
+    atomicAdd(&error, 1);
+  }
+}
+
+template <typename T>
+__global__ void test_multi_reduction_k(T* data, MultiReductionTestParams* p,
+                                       int* error) {
+  int t = blockIdx.x;
```

Comment:
`p[t]` and `error[t]` occur multiple times in the subsequent code. Consider loading them once (in addition to passing `data` into `test_single_radix()`):

```
MultiReductionTestParams params = p[blockIdx.x];
int* p_error = error[blockIdx.x];
switch (params.radix) {
  case 2:
    test_single_radix<2>(data, params, p_error);
    break;
  case 3:
    // ...
}
```